### DISCUSSION TITLE
How to delay or prevent Sparkle updates during internal app updates?

### DISCUSSION THREAD

**Discussion by honghoker** [sparkle-project/Sparkle#2669] [Q&A] [answered]
Hi! 👋
First of all, thanks to the Sparkle team for this great framework that I've been using successfully for app updates. 
I'm not sure if this is possible, but I have a question about managing Sparkle updates during internal app updates. 

I have a situation where my app performs internal updates, and I need to handle Sparkle updates appropriately during this process. Specifically:

1. When internal updates are running:
   - Either prevent the user from installing a Sparkle update even if they click "Install"
   - Or queue the Sparkle update to install after internal updates complete

2. Current issue:
   - Even when using SPUUpdaterDelegate, I can't find a way to prevent installation after the user clicks "Install" (when SPUUserUpdateChoice is .install)
   - The update proceeds regardless of my app's internal state

Questions:
1. Is there a way to intercept and delay/cancel Sparkle installations after the user clicks "Install"?
2. Alternatively, can I prevent the update dialog from appearing until internal updates complete?

I've tried working with SPUUpdaterDelegate methods but haven't found a solution that gives me this level of control over the update timing.

Any guidance or working examples would be greatly appreciated.

**Comment by zorgiepoo** [comment, marked as answer]
The install button sends an Apple quit event to the app. There are ways via NSApplicationDelegate to cancel or delay quit (`-[NSApplicationDelegate applicationShouldTerminate:]`). In your case you could delay quit until your internal process is done perhaps. If you want to prevent installation you may also want to delay ordinary user quit, maybe, depends.

`-[SPUUpdaterDelegate updater:mayPerformUpdateCheck:error:]` is for allowing an update check to be done (at the beginning). Potentially you could disallow this if your internal update is running.

`-[SPUUpdater sessionInProgress]` will tell if you if an update session is running. Potentially you could delay your internal update process if this is running.

There is also `-[SPUUpdaterDelegate updater:shouldPostponeRelaunchForUpdate:]` for delaying Install & Relaunch but I'm not sure if it works well. Worth exploring though.

**Reply by honghoker** [reply]
Thank you for your helpful response! 
Using - [SPUUpdaterDelegate updater:shouldPostponeRelaunchForUpdate:] during Install & Relaunch allowed me to delay and proceed with my tasks. 
I really appreciate your guidance
